src/app/public/modules/dynamic-component/dynamic-component.service.ts

```diff
@@ -0,0 +1,80 @@
+//#region imports
+
+import {
+  ApplicationRef,
+  ComponentFactoryResolver,
+  ComponentRef,
+  EmbeddedViewRef,
+  Injectable,
+  Injector,
+  RendererFactory2,
+  Renderer2
+} from '@angular/core';
+
+import {
+  SkyWindowRefService
+} from '../window';
+
+import {
+  SkyDynamicComponentArgs
+} from './dynamic-component-args';
+
+import {
+  SkyDynamicComponentLocation
+} from './dynamic-component-location';
+
+//#endregion
+
+/**
+ * Angular service for creating and rendering a dynamic component.
+ */
+@Injectable()
+export class SkyDynamicComponentService {
+
+  private renderer: Renderer2;
+
+  constructor(
+    private componentFactoryResolver: ComponentFactoryResolver,
+    private appRef: ApplicationRef,
+    private injector: Injector,
+    private windowRef: SkyWindowRefService,
+    rendererFactory: RendererFactory2
+  ) {
+    // Based on suggestions from https://github.com/angular/angular/issues/17824
+    // for accessing an instance of Renderer2 in a service since Renderer2 can't
+    // be injected into a service.  Passing undefined for both parameters results
+    // in the default renderer which is what we want here.
+    this.renderer = rendererFactory.createRenderer(undefined, undefined);
+  }
+
+  /**
+   * Creates an instance of the specified component and adds it to the specified location
+   * on the page.
+   * @param args Options for creating the dynamic component.
+   */
+  public createComponent<T>(args: SkyDynamicComponentArgs): ComponentRef<T> {
+    const cmpRef = this.componentFactoryResolver
+      .resolveComponentFactory<T>(args.componentType)
+      .create(this.injector);
+
+    this.appRef.attachView(cmpRef.hostView);
```

Comment:
Flyout is a good example of two different types of dynamic component injection: 1) injecting a component outside of the app's markup (ie, the `body` tag), or 2) injecting a component into another `ViewContainerRef`.

Should we accommodate both use cases (or at least rename this method with a future update in mind)?

Another aspect to consider is allowing for custom providers to be configured (the second reference below demonstrates this).

Flyout injecting a component at the end of the `body` tag:
https://github.com/blackbaud/skyux-flyout/blob/master/src/app/public/modules/flyout/flyout.service.ts#L87-L98

Flyout injecting a component into an existing `ViewContainerRef`:
https://github.com/blackbaud/skyux-flyout/blob/master/src/app/public/modules/flyout/flyout.component.ts#L170-L173

Reply:
Makes sense. So in the (future) case that we wanted to inject a component inside a `ViewContainerRef`, would we create a new `location` type, like this?

```
this.dynamicComponentService.createComponent({
  componentType: MyComponent,
  location: SkyDynamicComponentLocation.Custom // or Target, or...?
});
```
Just trying to wrap my head around the API.

Reply:
Yes, the idea is that we would add a new location type and a new option to specify the `ViewContainerRef`.